Security Context Update

mew2057 opened the pull request:
The following is a Pull  Request to enable the namespaced operator on OpenShift:

- Removes Privilege mode from driver container.
- Adds an `scc.yaml` file to deploy.
- Updates  `README.md` to include `scc.yaml`.

@yadaven

yadaven commented on stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator/deploy/scc.yaml:
```diff
@@ -0,0 +1,45 @@
+supplementalGroups:
+  type: RunAsAny
+users:
+- system:serviceaccount:$namespace:ibm-spectrum-scale-csi-attacher
```
during my testing I changed $namespace to ibm-spectrum-scale-csi-driver namespace. Forgot to update during creation of patch. We should change $namespace -> ibm-spectrum-scale-csi-driver.

yadaven commented on stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator/deploy/scc.yaml:
```diff
@@ -0,0 +1,45 @@
+allowHostDirVolumePlugin: true
```
this SCC has been created by modifying restricted scc (https://docs.openshift.com/enterprise/3.0/admin_guide/manage_scc.html#examining-a-security-context-constraints-object)  It could be created as below
kind: SecurityContextConstraints
apiVersion: v1
metadata:
  name: scc-admin

But another way to create is take already present SCC and modify based on need.

yadaven replied:
I agree we have one yaml file and step for user to install operator. I am not sure how it will work with OLM ? does OLM will create its own SCC ?  Can you try this with OLM and let me know if creation of SCC is required there ?

yadaven replied:
OR should we merge some of the yaml files together, like SA, role, role_binding can be single yaml file

mew2057 commented:
Created a build off of this PR: quay.io/ibm-spectrum-scale/ibm-spectrum-scale-csi-operator:v0.9.2

mew2057 commented:
![Screen Shot 2019-11-19 at 2 31 41 PM](https://user-images.githubusercontent.com/1195452/69179382-70827400-0ad9-11ea-8053-bd6b56b1a786.png)

It looks like olm doesn't like the privilege the attacher has when deploying on openshift.

mew2057 commented:
This is problematic, because it essentially breaks OperatorHub compatibility.

yadaven commented:
> This is problematic, because it essentially breaks OperatorHub compatibility.



> ![Screen Shot 2019-11-19 at 2 31 41 PM](https://user-images.githubusercontent.com/1195452/69179382-70827400-0ad9-11ea-8053-bd6b56b1a786.png)
> 
> It looks like olm doesn't like the privilege the attacher has when deploying on openshift.

This error is coming because we have not added SCC.  With OLM can we add our sample scc.yaml file, I think with that this issue will go away

mew2057 commented:
We have two files to work with in OLM the CSV and the CRD. Additionally, the user accounts have hard coded namespaces. We need to integrate the scc, yes, but we need to figure out  how to do it turn key.

mew2057 commented:
I moved the `scc.yaml` into the operator for authorization. I'm not sure if this is the solution, because my subscription is not updating in my OpenShift environment.

Because we're calling out specific user accounts I'm concerned that this might  clobber  other namespaces (A no no in operator land).

This is the error code I was getting, I've already added that permission to the operator's roles, so we still need to see a run with  the updated manifest.

```
"Failed to retrieve requested object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"securitycontextconstraints.security.openshift.io \\\\\"csiaccess\\\\\" is forbidden: User \\\\\"system:serviceaccount:marketplace:ibm-spectrum-scale-csi-operator\\\\\" cannot get resource \\\\\"securitycontextconstraints\\\\\" in API group \\\\\"security.openshift.io\\\\\" at the cluster scope\",\"reason\":\"Forbidden\",\"details\":{\"name\":\"csiaccess\",\"group\":\"security.openshift.io\",\"kind\":\"securitycontextconstraints\"},\"code\":403}\\n'",
```

mew2057 commented:
Might roll this back tomorrow, because it looks like both Rook and the [Open Liberty](https://operatorhub.io/operator/open-liberty) operators use SCCs in their OperatorHub  deployments, which requires a manual step from the admin.

Still I'm worried about the operator hub  entry not being as turnkey as possible.

mew2057 commented:
Then again [node-problem-detector](https://github.com/operator-framework/community-operators/blob/10a676d6f1e1cc02f16c93c0ec6b20e4d22d02b1/community-operators/node-problem-detector/0.0.1/node-problem-detector.v0.0.1.clusterserviceversion.yaml) looks like it might do something like my last commit, since it has a  call out for `securitycontextconstraints` access.

whowutwut commented on stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator/deploy/olm-catalog/ibm-spectrum-scale-csi-operator/0.9.1/ibm-spectrum-scale-csi-operator.v0.9.1.clusterserviceversion.yaml:
```diff
@@ -18,7 +18,7 @@ metadata:
     categories: Storage
     certified: 'false'
-    containerImage: quay.io/ibm-spectrum-scale/ibm-spectrum-scale-csi-operator:v0.9.1
+    containerImage: quay.io/ibm-spectrum-scale/ibm-spectrum-scale-csi-operator:v0.9.2
```
This file has a PATH with  `0.9.1`  in the directory and in the name of this yaml file...  but in this file, we are changing to point to image `0.9.2`,  this does not seem correct

mew2057 replied:
I agree, the CSV needs to be pushed to  `0.9.2`.

whowutwut replied:
so we need to create a  file for each version bump?  How would this work if we had daily  builds going  to quay?

mew2057 commented:
@whowutwut This should now work on OpenShift, I did manual testing in cluster2, but it might be  better to run it in a cluster with scale actually backing it.

This  is the special sauce by the way:
```YAML
- name: "Ensure SCC"
  when: '"apps.openshift.io" in api_groups'
  block:
  - name: Ensure secret {{ item.secrets }} defined in {{ lookup( 'vars','namespace' ) }}
    k8s_facts:
      kind:  "SecurityContextConstraints"
      name : "csiaccess"
    register: csiaccess_results

  - name: "Ensure the csiaccess_users list"
    set_fact:
      csiaccess_users: "{{ (csiaccess_results.resources[0].users | default([])) }}"
      scc_state: "present"
      csiaccess_users_new:
        - 'system:serviceaccount:{{ namespace }}:ibm-spectrum-scale-csi-attacher'
        - 'system:serviceaccount:{{ namespace }}:ibm-spectrum-scale-csi-provisioner'
        - 'system:serviceaccount:{{ namespace }}:ibm-spectrum-scale-csi-node'

  - name: "Union users if state is present"
    set_fact:
      csiaccess_users: "{{csiaccess_users | union(csiaccess_users_new) }}"
    when: 'state == "present"'

  - name: "Difference users if state is absent"
    set_fact:
      csiaccess_users:  "{{csiaccess_users | difference(csiaccess_users_new)}}"
    when: 'state != "present"'

  - name: "Update scc_state"
    set_fact:
      scc_state: "absent"
    when: '(csiaccess_users|length) == 0'

  - name: "Ensure SCC are {{ state }}"
    k8s:
      state: "{{scc_state}}"
      namespace: "{{namespace}}"
      definition: "{{ lookup('template', 'scc.yaml.j2') | from_yaml }}"
```

Basically I templatized the `scc.yaml` we had and moved ownership  to the operator. This does three things for us:
1. Ensures the existing users are  preserved when updating `csiaccess` 
2. If finalizers are enabled, ensures the `scc` record for `csiaccess` gets cleared (I turned them back on and added a debug to the README).
3. Improves the level of "turn-key"  support for the operator in OperatorHub and OLM.
---

I think this is clear to merge feature-wise, I just want to see some test outside of me.

mew2057 commented:
Verified this is clean for a `kubernetes` install, and as far as I know `OpenShift` seems to be working as well.

yadaven commented:
> @whowutwut This should now work on OpenShift, I did manual testing in cluster2, but it might be better to run it in a cluster with scale actually backing it.
> 
> This is the special sauce by the way:
> 
> ```yaml
> - name: "Ensure SCC"
>   when: '"apps.openshift.io" in api_groups'
>   block:
>   - name: Ensure secret {{ item.secrets }} defined in {{ lookup( 'vars','namespace' ) }}
>     k8s_facts:
>       kind:  "SecurityContextConstraints"
>       name : "csiaccess"
>     register: csiaccess_results
> 
>   - name: "Ensure the csiaccess_users list"
>     set_fact:
>       csiaccess_users: "{{ (csiaccess_results.resources[0].users | default([])) }}"
>       scc_state: "present"
>       csiaccess_users_new:
>         - 'system:serviceaccount:{{ namespace }}:ibm-spectrum-scale-csi-attacher'
>         - 'system:serviceaccount:{{ namespace }}:ibm-spectrum-scale-csi-provisioner'
>         - 'system:serviceaccount:{{ namespace }}:ibm-spectrum-scale-csi-node'
> 
>   - name: "Union users if state is present"
>     set_fact:
>       csiaccess_users: "{{csiaccess_users | union(csiaccess_users_new) }}"
>     when: 'state == "present"'
> 
>   - name: "Difference users if state is absent"
>     set_fact:
>       csiaccess_users:  "{{csiaccess_users | difference(csiaccess_users_new)}}"
>     when: 'state != "present"'
> 
>   - name: "Update scc_state"
>     set_fact:
>       scc_state: "absent"
>     when: '(csiaccess_users|length) == 0'
> 
>   - name: "Ensure SCC are {{ state }}"
>     k8s:
>       state: "{{scc_state}}"
>       namespace: "{{namespace}}"
>       definition: "{{ lookup('template', 'scc.yaml.j2') | from_yaml }}"
> ```
> 
> Basically I templatized the `scc.yaml` we had and moved ownership to the operator. This does three things for us:
> 
>     1. Ensures the existing users are  preserved when updating `csiaccess`
> 
>     2. If finalizers are enabled, ensures the `scc` record for `csiaccess` gets cleared (I turned them back on and added a debug to the README).
> 
>     3. Improves the level of "turn-key"  support for the operator in OperatorHub and OLM.
> 
> 
> I think this is clear to merge feature-wise, I just want to see some test outside of me.

Had look at changes, it looks good, will try one round of testing on OpenShift and K8s and let you know the feedback.

whowutwut commented:
I will also try and build/test  this tomorrow.

whowutwut commented:
Ok, was quicker to test than I had thought...

Changed the playbook to point at this branch... 
```
# repo_user: "IBM"
# repo_branch: "v0.9.0"
repo_user: "mew2057"
repo_branch: "security"
```

Re-ran the playbook... and then deploy Operator... 

```
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh deployOP
No resources found.
namespace/ibm-spectrum-scale-csi-driver created
serviceaccount/ibm-spectrum-scale-csi-operator created
serviceaccount/ibm-spectrum-scale-csi-attacher created
serviceaccount/ibm-spectrum-scale-csi-node created
serviceaccount/ibm-spectrum-scale-csi-provisioner created
role.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-node created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-attacher created
clusterrole.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-provisioner created
rolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-operator created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-node created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-provisioner created
clusterrolebinding.rbac.authorization.k8s.io/ibm-spectrum-scale-csi-attacher created
customresourcedefinition.apiextensions.k8s.io/csiscaleoperators.scale.ibm.com created
deployment.apps/ibm-spectrum-scale-csi-operator created
No resources found.
Wed Nov 20 16:44:25 EST 2019: Waiting up to 15*40 seconds for the operator to come up ... [loop: 1]
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   0/2     ContainerCreating   0          0s
--- Running! ---
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running   0          16s
secret/spectrum-scale-gui-secret created
```

Then deploy driver: 
```
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh deployDR
csiscaleoperator.scale.ibm.com/ibm-spectrum-scale-csi created
[root@c943f4n01-pvt csi-operator-ansible]#
...
...
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running   0          38s
+ set +x
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-attacher-0                  0/1     ContainerCreating   0          2s
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running             0          41s
ibm-spectrum-scale-csi-provisioner-0               0/1     ContainerCreating   0          1s
+ set +x
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-attacher-0                  0/1     ContainerCreating   0          5s
ibm-spectrum-scale-csi-hq59t                       0/2     ContainerCreating   0          2s
ibm-spectrum-scale-csi-hrzh9                       0/2     ContainerCreating   0          2s
ibm-spectrum-scale-csi-k97hs                       0/2     ContainerCreating   0          2s
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running             0          44s
ibm-spectrum-scale-csi-provisioner-0               0/1     ContainerCreating   0          4s
+ set +x
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-attacher-0                  0/1     ContainerCreating   0          8s
ibm-spectrum-scale-csi-hq59t                       0/2     ContainerCreating   0          5s
ibm-spectrum-scale-csi-hrzh9                       0/2     ContainerCreating   0          5s
ibm-spectrum-scale-csi-k97hs                       0/2     ContainerCreating   0          5s
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running             0          47s
ibm-spectrum-scale-csi-provisioner-0               0/1     ContainerCreating   0          7s
+ set +x
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS              RESTARTS   AGE
ibm-spectrum-scale-csi-attacher-0                  0/1     ContainerCreating   0          11s
ibm-spectrum-scale-csi-hq59t                       0/2     ContainerCreating   0          8s
ibm-spectrum-scale-csi-hrzh9                       0/2     ContainerCreating   0          8s
ibm-spectrum-scale-csi-k97hs                       0/2     ContainerCreating   0          8s
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running             0          50s
ibm-spectrum-scale-csi-provisioner-0               0/1     ContainerCreating   0          10s
+ set +x
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-attacher-0                  1/1     Running   0          15s
ibm-spectrum-scale-csi-hq59t                       2/2     Running   0          12s
ibm-spectrum-scale-csi-hrzh9                       2/2     Running   0          12s
ibm-spectrum-scale-csi-k97hs                       2/2     Running   0          12s
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running   0          54s
ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          14s
+ set +x
[root@c943f4n01-pvt csi-operator-ansible]#
```

So at least a sanity bring up works...

whowutwut commented:
Checking back on the pods:
```
[root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
+ oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-8t9k2                       2/2     Running   0          29m
ibm-spectrum-scale-csi-attacher-0                  1/1     Running   0          4h53m
ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running   0          4h53m
ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          4h53m
ibm-spectrum-scale-csi-tcml4                       2/2     Running   0          29m
ibm-spectrum-scale-csi-x8nf6                       2/2     Running   0          29m
+ set +x
```

Why are the daemonSets have a much shorter AGE than the StatefulSets?

yadaven commented:
> Checking back on the pods:
> 
> ```
> [root@c943f4n01-pvt csi-operator-ansible]# ./operator-helper.sh get_pods
> + oc get pods -n ibm-spectrum-scale-csi-driver
> NAME                                               READY   STATUS    RESTARTS   AGE
> ibm-spectrum-scale-csi-8t9k2                       2/2     Running   0          29m
> ibm-spectrum-scale-csi-attacher-0                  1/1     Running   0          4h53m
> ibm-spectrum-scale-csi-operator-6ff9cf6979-k2gpd   2/2     Running   0          4h53m
> ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          4h53m
> ibm-spectrum-scale-csi-tcml4                       2/2     Running   0          29m
> ibm-spectrum-scale-csi-x8nf6                       2/2     Running   0          29m
> + set +x
> ```
> 
> Why are the daemonSets have a much shorter AGE than the StatefulSets?

There is some issue for sure. I see below logs in describe o/p of daemonset

 Type    Reason            Age   From                  Message
  ----    ------            ----  ----                  -------
  Normal  SuccessfulDelete  165m  daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-72pcn
  Normal  SuccessfulCreate  165m  daemonset-controller  Created pod: ibm-spectrum-scale-csi-9pdk2
  Normal  SuccessfulDelete  164m  daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-tqht6
  Normal  SuccessfulCreate  164m  daemonset-controller  Created pod: ibm-spectrum-scale-csi-45s9c
  Normal  SuccessfulDelete  164m  daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-tp4l5
  Normal  SuccessfulCreate  164m  daemonset-controller  Created pod: ibm-spectrum-scale-csi-slg2k
  Normal  SuccessfulDelete  164m  daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-dkfn9
  Normal  SuccessfulCreate  164m  daemonset-controller  Created pod: ibm-spectrum-scale-csi-4gp8z
  Normal  SuccessfulDelete  27m   daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-45s9c
  Normal  SuccessfulCreate  27m   daemonset-controller  Created pod: ibm-spectrum-scale-csi-bz4rp
  Normal  SuccessfulDelete  27m   daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-9pdk2
  Normal  SuccessfulCreate  27m   daemonset-controller  Created pod: ibm-spectrum-scale-csi-ct6nr
  Normal  SuccessfulDelete  26m   daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-slg2k
  Normal  SuccessfulCreate  26m   daemonset-controller  Created pod: ibm-spectrum-scale-csi-c85kc
  Normal  SuccessfulDelete  26m   daemonset-controller  Deleted pod: ibm-spectrum-scale-csi-4gp8z
  Normal  SuccessfulCreate  26m   daemonset-controller  Created pod: ibm-spectrum-scale-csi-swsv9
[root@drogon-inf log]# 

So we can see delete/create is happening after certain time.  Trying to debug it, didn;t find any specific failure logs from plugin side. digging more

mew2057 commented:
We might want to examine the state of the  pods, typically we can get  some insight from the `status` field.

mew2057 commented:
Looks like there's a 16 hour window between the first and last timestamp in the operator logs?

1574344980
1574286274

mew2057 commented:
It gets more interesting, in my raw kubernetes environment I'm getting this:
```
kubectl get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-attacher-0                  1/1     Running   0          17h
ibm-spectrum-scale-csi-h5pxd                       2/2     Running   0          7h18m
ibm-spectrum-scale-csi-operator-5db6fc7756-4qzm9   2/2     Running   0          17h
ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          17h
ibm-spectrum-scale-csi-zxtgs                       2/2     Running   0          7h18m
```

yadaven commented:
> It gets more interesting, in my raw kubernetes environment I'm getting this:
> 
> ```
> kubectl get pods -n ibm-spectrum-scale-csi-driver
> NAME                                               READY   STATUS    RESTARTS   AGE
> ibm-spectrum-scale-csi-attacher-0                  1/1     Running   0          17h
> ibm-spectrum-scale-csi-h5pxd                       2/2     Running   0          7h18m
> ibm-spectrum-scale-csi-operator-5db6fc7756-4qzm9   2/2     Running   0          17h
> ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          17h
> ibm-spectrum-scale-csi-zxtgs                       2/2     Running   0          7h18m
> ```

On worker nodes I get below logs

status for "ibm-spectrum-scale-csi-4gp8z_ibm-spectrum-scale-csi-driver(741d59b3-0c25-11ea-8972-00505695e251)"
messages:Nov 21 01:26:49 drogon-compute2 hyperkube: I1121 01:26:49.973071    2882 event.go:209] Event(v1.ObjectReference{Kind:"Pod", Namespace:"ibm-spectrum-scale-csi-driver", Name:"ibm-spectrum-scale-csi-4gp8z", UID:"741d59b3-0c25-11ea-8972-00505695e251", APIVersion:"v1", ResourceVersion:"4033346", FieldPath:"spec.containers{ibm-spectrum-scale-csi}"}): type: 'Normal' reason: 'Killing' Stopping container ibm-spectrum-scale-csi
messages:Nov 21 01:26:49 drogon-compute2 hyperkube: I1121 01:26:49.973116    2882 event.go:209] Event(v1.ObjectReference{Kind:"Pod", Namespace:"ibm-spectrum-scale-csi-driv

So someone is stopping the csi pods. I see kubelet has requested for DELETE, but who sends request to kubelet. Did we see same behaviour when our daemonset was privileged ?  I think I need to try that. If you have older dev branch changes running on some cluster can we check if same behavior was with older build i.e before scc was added ?

mew2057 commented:
I think  something is triggering the secret trigger, which is restarting the daemonset because the counter is getting bumped. I'm merging SCC because I don't think  this  code is at all related to the issue.
